cpp/src/traversal/mg/frontier_expand.cuh

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <graph.hpp>
+#include "frontier_expand_kernels.cuh"
+#include "vertex_binning.cuh"
+
+namespace cugraph {
+
+namespace mg {
+
+namespace detail {
+
+template <typename VT, typename ET, typename WT>
+class FrontierExpand {
+  raft::handle_t const &handle_;
+  cugraph::GraphCSRView<VT, ET, WT> const &graph_;
+  VertexBinner<VT, ET> dist_;
+  rmm::device_vector<VT> reorganized_vertices_;
+  ET vertex_begin_;
+  ET vertex_end_;
+  rmm::device_vector<ET> output_vertex_count_;
+
+ public:
+  FrontierExpand(raft::handle_t const &handle, cugraph::GraphCSRView<VT, ET, WT> const &graph)
+    : handle_(handle), graph_(graph)
+  {
+    bool is_mg = (handle.comms_initialized() && (graph.local_vertices != nullptr) &&
+                  (graph.local_offsets != nullptr));
+    if (is_mg) {
+      reorganized_vertices_.resize(graph.local_vertices[handle_.get_comms().get_rank()]);
+      vertex_begin_ = graph.local_offsets[handle_.get_comms().get_rank()];
+      vertex_end_   = graph.local_offsets[handle_.get_comms().get_rank()] +
+                    graph.local_vertices[handle_.get_comms().get_rank()];
+    } else {
+      reorganized_vertices_.resize(graph.number_of_vertices);
+      vertex_begin_ = 0;
+      vertex_end_   = graph.number_of_vertices;
+    }
+    output_vertex_count_.resize(1);
+  }
+
+  // Return the size of the output_frontier
+  template <typename Operator>
+  VT run(Operator op,
```

Comment:
Rather than a class with a `run()` method, perhaps overload `operator()` and make this class a functor?